File: BowlingGame/ruby/Simple/bowling_game_test.rb
$:.unshift File.dirname(__FILE__)

require "minitest/unit"
require "minitest/autorun"
require 'bowling_game'

class BowlingGameTest < MiniTest::Unit::TestCase

  def setup
    @game = BowlingGame.new
  end

  def test_gutter_game
    roll_many 20, 0
    assert_equal 0, @game.score
  end

  def test_roll_all_ones
    roll_many 20, 1
    assert_equal 20, @game.score
  end

  def test_roll_spare
    roll_spare
    roll_many 17, 0

    assert_equal 16, @game.score
  end

  def test_roll_strike
    roll_strike
    roll_many 16, 0

    assert_equal 24, @game.score
  end

  def test_roll_perfect_game
    roll_many 12, 10
    assert_equal 300, @game.score
  end

  def roll_many(num, pins_hit)
    num.times { @game.roll pins_hit}
  end

  def roll_strike
    @game.roll 10, 3, 4
  end

  def roll_spare
    @game.roll 5, 5, 3
  end
end

File: BowlingGame/java/Simple/src/BowlingGameTest.java
import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;

public class BowlingGameTest {

    private Game game;

    @Before
    public void setUp() throws Exception {
        game = new Game();
    }

    @Test
    public void gutterGameTest() {
        rollMany(20, 0);

        assertEquals(0, game.score());
    }

    private void rollMany(int numRolls, int pinsDown) {
        for (int i = 0; i<numRolls; i++)
            game.roll(pinsDown);
    }

    @Test
    public void rollAllOnesTest() {
        rollMany(20, 1);

        assertEquals(20, game.score());
    }

    @Test
    public void rollOneSpare() {
        rollSpare();

        game.roll(3);
        game.roll(0);

        rollMany(16, 0);

        assertEquals(16, game.score());
    }

    private void rollSpare() {
        game.roll(5);
        game.roll(5);
    }

    @Test
    public void rollOneStrikeTest() {
        rollStrike();

        game.roll(3);
        game.roll(4);

        rollMany(17, 0);

        assertEquals(24, game.score());
    }

    private void rollStrike() {
        game.roll(10);
    }

    @Test
    public void rollPerfectGameTest() {
        rollMany(12, 10);

        assertEquals(300, game.score());
    }
}
